### PR TITLE
Simplify negate() implementation

### PR DESCRIPTION
**Pull request by lionel-**
I think the general strategy should look like this. For primitive functions, I wonder if `as_function()` should create a closure wrapper around them? E.g.

```r
as_function(`-`)
#> function(e1, e2) `-`(e1, e2)
```

This makes sure argument matching behaves properly.

### CODE REVIEW

**Comment by lionel-** [comment]
@hadley I think `as_mapper.default()` should forward to `rlang::as_function()`. Does that sound reasonable to you? Then we'd move the primitive handling there.

File: R/negate.R [hunk 1]
```diff
@@ -14,33 +14,11 @@
...
-    args <- alist(x = )
-    body <- lang_interp(extract(x, UQ(as.list(f)), default = UQ(default)))
-  } else {
-    stop("Unknown input", call. = FALSE)
-  }
-
-  new_fn(args, trans(body), f_env(quo_f))
-}
+  body(.p) <- tidy_interp(quote({
+    ! { !!! body(.p) }
```
Comment (hadley) [inline R/negate.R]: Shouldn't this be `()`?

**Comment by hadley** [comment]
Ah, this is much simpler!  I agree that `as_mapper.default()` should just call `rlang::as_function()`

File: R/negate.R (hunk 1 above)
Comment (lionel-) [inline R/negate.R]: oh right, the body will include the curlies if it has several statements

**Comment by lionel-** [comment]
`as_mapper()` now forwards to `as_function()` for both formulas and functions. I wonder if both of these should be turned to switchpatches though?

**Comment by hadley** [comment]
Switchpatch doesn't feel right to me here. `as_function()` could do switchpatch, but I don't think `as_mapper()` should.

**Pull request merged**